ChildProject/annotations.py

```diff
@@ -122,6 +130,17 @@ def load_textgrid(self, filename):
         path = os.path.join(self.project.path, 'raw_annotations', filename)
         textgrid = pympi.Praat.TextGrid(path)
 
+        def ling_type(s):
```

Comment:
is this good for you ? @alecristia

Reply:
interesting syntax! does this mean:

- check if s is a string (and NOT convert into a string)
- if false, then return NA

What may or may not be a string? In r this depends on how you read something: in read.table, fields that have only numbers will be some form of numeric; otherwise they'll be factors -- so nothing would be string.

Examples of human coding errors:

- "1" (not a string, right? if yes, handled by l. 138-10)
- "0" (idem, both are correct)
-  " 0" (error entering it, but we get what coder meant - "0"; and this would be a string so not NAd?)
-  "¨0" (error entering it, but we get what coder meant - "0"; and this would be a string so not NAd?)
- "00" (error entering it, but we get what coder meant - "0"; and this is handled by lines 138-142 - to which I'll return)
- "x" (impossible to tell what coder meant ; and this is handled by lines 138-142)
- "01" (impossible to tell what coder meant ; and this is not yet handled by lines 138-142)

--> suggested correction: substitute out all non [0-9], so that only [0-9] remain in s

lines 138-142:

- define two boolean vars, a & b, which check whether s contains 0 or 1 -- note: in R, when a variable is FALSE, all matches are accepted (ie FALSE contains 0) - I'm guessing this doesn't happen in py?
- if either a or b are true then return a 0 -- return 0 if a is true, and otherwise 1
- BUT this means that "01" gets a "0", but we can't be sure -- so can we make a^b disjunctive? or add a line so that when both are true s in NA?

Reply:
mmh, i think the code is doing what you're saying ?
Here is an example 👍 
```python
def ling_type(s):
    s = str(s)
    if not s:
        return "NA"

    a, b = ('0' in s, '1' in s)
    if a^b:
        return '0' if a else '1' 
    else:
        return 'NA'

values = ['1', '0', ' 0', '¨0', 'x', '01', '', 1]

print("\n".join([
    "ling_type('{}') = {}".format(s, ling_type(s))
    for s in values
]))
```

Output:
```
ling_type('1') = 1
ling_type('0') = 0
ling_type(' 0') = 0
ling_type('¨0') = 0
ling_type('x') = NA
ling_type('01') = NA
ling_type('') = NA
ling_type('1') = 1
```

and I've used a^b because it is disjoint indeed:

| a | b | a^b |
|---|---|-----|
| 0 | 0 | 0   |
| 0 | 1 | 1   |
| 1 | 0 | 1   |
| 1 | 1 | 0   |

so that ambiguous values such as 01 return NA :)

Reply:
perfect!